src/Covenant/Internal/Rename.hs

```diff
@@ -201,6 +204,32 @@ renameValT = \case
   Abstraction t -> Abstraction <$> renameAbstraction t
   ThunkT t -> ThunkT <$> renameCompT t
   BuiltinFlat t -> pure . BuiltinFlat $ t
+  -- Assumes kind-checking has occurred
+  Datatype tn xs -> RenameM $ do
+    -- We don't step or un-step the scope here b/c a TyCon which appears as a ValT _cannot_ bind variables.
+    -- This Vector here doesn't represent a function, but a product, so we there is no "return" type to treat specially (I think!)
+    renamedXS <-
+      Vector.generateM
+        (Vector.length xs)
+        (\i -> coerce . renameValT $ xs Vector.! i)
+    pure $ Datatype tn renamedXS
+
+-- @since 1.1.0
+renameDataDecl :: DataDeclaration AbstractTy -> RenameM (DataDeclaration Renamed)
+renameDataDecl (DataDeclaration tn cnt ctors) = RenameM $ do
+  modify (stepUpScope cnt)
+  renamedCtors <-
+    Vector.generateM
+      (Vector.length ctors)
+      (\i -> coerce . renameCtor $ ctors Vector.! i)
+  -- REVIEW: @Koz is it ok to skip this here? It SEEMS ok
+  -- ourAbstractions <- gets (view (#tracker % to Vector.head % _1))
+  -- unless (Vector.and ourAbstractions) (throwError $ UndeterminedAbstraction)
```

Comment:
I think you have to. The determinism checker can and will vomit on data types which have phantom 'arms'.